system-addon/lib/SectionsManager.jsm

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {utils: Cu} = Components;
+const {actionCreators: ac, actionTypes: at} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+
+this.SectionsFeed = class SectionsFeed {
+  constructor() {
+    this.onAddSection = this.onAddSection.bind(this);
+    this.onRemoveSection = this.onRemoveSection.bind(this);
+    this.onUpdateRows = this.onUpdateRows.bind(this);
+  }
+
+  init() {
+    SectionsManager.addListener(SectionsManager.ADD_SECTION, this.onAddSection);
+    SectionsManager.addListener(SectionsManager.REMOVE_SECTION, this.onRemoveSection);
+    SectionsManager.addListener(SectionsManager.UPDATE_ROWS, this.onUpdateRows);
+    // Run for any already existing sections
+    for (const id of SectionsManager.sections.keys()) {
+      this.onAddSection(id);
+    }
+  }
+
+  uninit() {
+    SectionsManager.removeListener(SectionsManager.ADD_SECTION, this.onAddSection);
+    SectionsManager.removeListener(SectionsManager.REMOVE_SECTION, this.onRemoveSection);
+    SectionsManager.removeListener(SectionsManager.UPDATE_ROWS, this.onUpdateRows);
+  }
+
+  onAddSection(id) {
+    const {options} = SectionsManager.sections.get(id);
+    if (options) {
+      this.store.dispatch(ac.BroadcastToContent({type: at.SECTION_REGISTER, data: Object.assign({id}, options)}));
+    }
+  }
+
+  onRemoveSection(id) {
+    this.store.dispatch(ac.BroadcastToContent({type: at.SECTION_DEREGISTER, data: id}));
+  }
+
+  onUpdateRows({id, shouldBroadcast}) {
+    const {rows} = SectionsManager.sections.get(id);
+    if (rows) {
+      const action = {type: at.SECTION_ROWS_UPDATE, data: Object.assign({id}, {rows})};
+      this.store.dispatch(shouldBroadcast ? ac.BroadcastToContent(action) : action);
+    }
+  }
+
+  onAction(action) {
+    switch (action.type) {
+      case at.INIT:
+        this.init();
+        break;
+    }
+    SectionsManager.notifyListeners(action.type);
+  }
+};
+
+this.SectionsManager = {
+  ADD_SECTION: "ADD_SECTION",
+  REMOVE_SECTION: "REMOVE_SECTION",
+  UPDATE_ROWS: "UPDATE_ROWS",
+  sections: new Map(),
+  listeners: new Map(),     // Subscribed to a specific event
+  feedListeners: new Set(), // Subscribed to all events
+  addListener(event, callback) {
```

Comment:
Would it be possible to use `resource://gre/modules/EventEmitter.jsm` instead of reimplementing this pattern?